File: 0200-number-of-islands/0200-number-of-islands.java
 class Solution {
  public int numIslands(char[][] grid) {
    int result = 0;
    for (int i = 0; i < grid.length; i++) {
      for (int j = 0; j < grid[0].length; j++) {
        if (grid[i][j] == '1') {
          dfs(grid, i, j);
          result++;
        }
      }
    }
  return result;
  }
  
  void dfs(char[][] grid, int x, int y) {
    if (x >= grid.length || x < 0 || y < 0 || y >= grid[x].length || grid[x][y] == '0') {
      return;
    }
    grid[x][y] = '0';
    dfs(grid, x + 1, y);
    dfs(grid, x - 1, y);
    dfs(grid, x, y + 1);
    dfs(grid, x, y - 1);
  }
}

File: 0200-number-of-islands/0200-number-of-islands.scala
object Solution {
    def numIslands(grid: Array[Array[Char]]): Int = {
        var result = 0
        for (i <- 0 to grid.length - 1) {
            for (j <- 0 to grid(0).length - 1) {
                if (grid(i)(j) == '1') {
                    result += 1
                    dfs(grid, i, j)
                }
               
            }
        }
        result
    }
    def dfs(grid: Array[Array[Char]], i: Int, j: Int): Unit = {
        if (i < 0 || j < 0 || j >= grid(0).length || i >= grid.length || grid(i)(j) == '0') {
            return
        } else {
            grid(i)(j) = '0'
            dfs(grid, i + 1, j)
            dfs(grid, i - 1, j)
            dfs(grid, i, j + 1)
            dfs(grid, i, j - 1)
        }
        
    }
    
}